File: client/src/relay/sse.py
"""SSE (Server-Sent Events) client for the WRAI.TH relay activity stream.

Maintains a persistent connection to /api/activity/stream and dispatches
events to registered handlers. Auto-reconnects with backoff on disconnect.
"""

from __future__ import annotations

import asyncio
import json
from collections.abc import Callable, Coroutine
from typing import Any

import httpx
import structlog

from .models import SSEEvent

log = structlog.get_logger()

EventHandler = Callable[[SSEEvent], Coroutine[Any, Any, None]]


class SSEClient:
    """Persistent SSE connection to the relay activity stream."""

    def __init__(
        self,
        base_url: str,
        project: str = "default",
        reconnect_delay: float = 3.0,
    ):
        self.base_url = base_url.rstrip("/")
        self.project = project
        self.reconnect_delay = reconnect_delay
        self._handlers: dict[str, list[EventHandler]] = {}
        self._running = False
        self._task: asyncio.Task | None = None

    def on(self, event_type: str, handler: EventHandler):
        """Register a handler for an event type.

        Event types: message, activity, agent_status, task_update, memory_conflict
        Use "*" to catch all events.
        """
        self._handlers.setdefault(event_type, []).append(handler)

    async def start(self):
        """Start the SSE connection loop."""
        self._running = True
        self._task = asyncio.create_task(self._loop())
        log.info("sse.started", url=self.base_url, project=self.project)

    async def stop(self):
        """Stop the SSE connection."""
        self._running = False
        if self._task:
            self._task.cancel()
            try:
                await self._task
            except asyncio.CancelledError:
                pass
        log.info("sse.stopped")

    async def _loop(self):
        """Main connection loop with auto-reconnect."""
        delay = self.reconnect_delay

        while self._running:
            try:
                await self._connect()
                delay = self.reconnect_delay  # reset on successful connection
            except httpx.ConnectError:
                log.warning("sse.connect_failed", delay=delay)
            except httpx.ReadError:
                log.warning("sse.read_error", delay=delay)
            except Exception as e:
                log.error("sse.error", error=str(e), delay=delay)

            if self._running:
                await asyncio.sleep(delay)
                delay = min(delay * 1.5, 60.0)  # backoff up to 60s

    async def _connect(self):
        """Connect to SSE stream and process events."""
        url = f"{self.base_url}/api/activity/stream?project={self.project}"
        log.info("sse.connecting", url=url)

        async with httpx.AsyncClient(timeout=None) as client:
            async with client.stream("GET", url) as resp:
                resp.raise_for_status()
                log.info("sse.connected")

                event_type = ""
                data_lines: list[str] = []

                async for line in resp.aiter_lines():
                    if not self._running:
                        return

                    line = line.strip()

                    if line.startswith("event:"):
                        event_type = line[6:].strip()
                    elif line.startswith("data:"):
                        data_lines.append(line[5:].strip())
                    elif line == "" and data_lines:
                        # End of event — dispatch
                        raw = "\n".join(data_lines)
                        data_lines = []
                        try:
                            data = json.loads(raw)
                            event = SSEEvent(
                                type=event_type or data.get("type", ""),
                                data=data,
                            )
                            await self._dispatch(event)
                        except json.JSONDecodeError:
                            log.warning("sse.invalid_json", raw=raw[:200])
                        event_type = ""

    async def _dispatch(self, event: SSEEvent):
        """Dispatch event to registered handlers."""
        handlers = self._handlers.get(event.type, []) + self._handlers.get("*", [])
        for handler in handlers:
            try:
                await handler(event)
            except Exception as e:
                log.error("sse.handler_error", type=event.type, error=str(e))

File: internal/client/sse.go
package client

import (
	"bufio"
	"context"
	"encoding/json"
	"fmt"
	"log"
	"net/http"
	"net/url"
	"strings"
	"time"
)

// SSEEvent represents a parsed event from the relay activity stream.
type SSEEvent struct {
	Type    string                 `json:"type"`    // message, task, activity, agent_status, memory, register, goal, conversation
	Action  string                 `json:"action"`  // e.g. dispatch, claim, set, conflict
	Agent   string                 `json:"agent"`   // source agent
	Project string                 `json:"project"`
	Target  string                 `json:"target"`  // target agent or profile
	Label   string                 `json:"label"`
	Data    map[string]interface{} `json:"data"`    // full event payload
	TS      int64                  `json:"ts"`
}

// EventHandler processes an SSE event.
type EventHandler func(SSEEvent)

// SSEClient maintains a persistent connection to the relay activity stream.
type SSEClient struct {
	url            string
	project        string
	apiKey         string
	reconnectDelay time.Duration
	handlers       map[string][]EventHandler
}

// NewSSEClient creates an SSE client.
func NewSSEClient(baseURL, project, apiKey string, reconnectDelay time.Duration) *SSEClient {
	return &SSEClient{
		url:            baseURL,
		project:        project,
		apiKey:         apiKey,
		reconnectDelay: reconnectDelay,
		handlers:       make(map[string][]EventHandler),
	}
}

// On registers a handler for an event type.
// Use "*" to handle all events.
func (s *SSEClient) On(eventType string, handler EventHandler) {
	s.handlers[eventType] = append(s.handlers[eventType], handler)
}

// Run connects to the SSE stream and dispatches events.
// Reconnects automatically on disconnect.
func (s *SSEClient) Run(ctx context.Context) {
	for {
		select {
		case <-ctx.Done():
			return
		default:
		}

		err := s.connect(ctx)
		if err != nil && ctx.Err() == nil {
			log.Printf("[sse] disconnected: %v, reconnecting in %v", err, s.reconnectDelay)
		}

		select {
		case <-ctx.Done():
			return
		case <-time.After(s.reconnectDelay):
		}
	}
}

func (s *SSEClient) connect(ctx context.Context) error {
	u := fmt.Sprintf("%s/api/activity/stream?project=%s", s.url, url.QueryEscape(s.project))

	req, err := http.NewRequestWithContext(ctx, "GET", u, nil)
	if err != nil {
		return fmt.Errorf("create request: %w", err)
	}
	req.Header.Set("Accept", "text/event-stream")
	if s.apiKey != "" {
		req.Header.Set("Authorization", "Bearer "+s.apiKey)
	}

	client := &http.Client{Timeout: 0} // no timeout for SSE
	resp, err := client.Do(req)
	if err != nil {
		return fmt.Errorf("connect: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return fmt.Errorf("HTTP %d", resp.StatusCode)
	}

	log.Printf("[sse] connected to %s", u)

	scanner := bufio.NewScanner(resp.Body)
	scanner.Buffer(make([]byte, 256*1024), 256*1024) // 256KB buffer

	var eventType string
	var dataLines []string

	for scanner.Scan() {
		line := scanner.Text()

		if line == "" {
			// Empty line = end of event
			if eventType != "" && len(dataLines) > 0 {
				data := strings.Join(dataLines, "\n")
				s.dispatch(eventType, data)
			}
			eventType = ""
			dataLines = nil
			continue
		}

		if strings.HasPrefix(line, "event:") {
			eventType = strings.TrimSpace(strings.TrimPrefix(line, "event:"))
		} else if strings.HasPrefix(line, "data:") {
			dataLines = append(dataLines, strings.TrimSpace(strings.TrimPrefix(line, "data:")))
		}
	}

	if err := scanner.Err(); err != nil {
		return fmt.Errorf("read stream: %w", err)
	}
	return fmt.Errorf("stream closed")
}

func (s *SSEClient) dispatch(eventType, data string) {
	var raw map[string]interface{}
	if err := json.Unmarshal([]byte(data), &raw); err != nil {
		log.Printf("[sse] invalid JSON in %s event: %v", eventType, err)
		return
	}

	evt := SSEEvent{
		Type: eventType,
		Data: raw,
	}

	// Extract common fields
	if v, ok := raw["action"].(string); ok {
		evt.Action = v
	}
	if v, ok := raw["agent"].(string); ok {
		evt.Agent = v
	}
	if v, ok := raw["project"].(string); ok {
		evt.Project = v
	}
	if v, ok := raw["target"].(string); ok {
		evt.Target = v
	}
	if v, ok := raw["label"].(string); ok {
		evt.Label = v
	}
	if v, ok := raw["ts"].(float64); ok {
		evt.TS = int64(v)
	}

	// Dispatch to typed handlers
	if handlers, ok := s.handlers[eventType]; ok {
		for _, h := range handlers {
			h(evt)
		}
	}

	// Dispatch to wildcard handlers
	if handlers, ok := s.handlers["*"]; ok {
		for _, h := range handlers {
			h(evt)
		}
	}
}
